sim/src/main/cc/bridges/TracerVModule.cc

```diff
@@ -0,0 +1,367 @@
+// See LICENSE for license details.
+
+#include "bridges/peek_poke.h"
+#include "bridges/tracerv.h"
+#include "core/bridge_driver.h"
+#include "core/simif.h"
+#include "core/simulation.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string_view>
+
+static std::vector<bool> get_contiguous(const unsigned bits,
+                                        const unsigned total) {
+  std::vector<bool> ret;
+  for (unsigned i = 0; i < total; i++) {
+    const bool value = (i < bits);
+    ret.emplace_back(value);
+  }
+
+  return ret;
+}
+
+static std::vector<uint64_t> get_iaddrs(const unsigned step,
+                                        const unsigned total) {
+  constexpr uint64_t offset =
+      1024; // should be larger than total but doesn't really matter
+  std::vector<uint64_t> ret;
+
+  for (unsigned i = 0; i < total; i++) {
+    ret.emplace_back(step * offset + i);
+  }
+
+  return ret;
+}
+
+static std::string namei(const unsigned x) {
+  std::stringstream ss;
+  ss << "io_insns_" << x << "_iaddr";
+  return ss.str();
+};
+
+static std::string nameinsn(const unsigned x) {
+  std::stringstream ss;
+  ss << "io_insns_" << x << "_insn";
+  return ss.str();
+};
+
+static std::string namev(const unsigned x) {
+  std::stringstream ss;
+  ss << "io_insns_" << x << "_valid";
+  return ss.str();
+};
+
+// a simple way to turn the iaddr values we use
+// in this test into a 5 bit value
+static uint32_t fold(const uint64_t x) {
+  uint32_t a = x & 0x1f;
+  uint32_t b = x >> 8;
+  uint32_t c = x >> 13;
+  return (a ^ b ^ c) & 0x1f;
+}
+
+class TracerVModule : public simulation_t {
+private:
+  peek_poke_t &peek_poke;
+  tracerv_t &tracerv;
+
+public:
+  TracerVModule(widget_registry_t &registry,
+                const std::vector<std::string> &args,
+                std::string_view target_name)
+      : simulation_t(registry, args),
+        peek_poke(registry.get_widget<peek_poke_t>()),
+        tracerv(registry.get_widget<tracerv_t>()) {
+
+    for (auto &arg : args) {
+      if (arg.find("+seed=") == 0) {
+        random_seed = strtoll(arg.c_str() + 6, nullptr, 10);
+        fprintf(stderr, "Using custom SEED: %ld\n", random_seed);
+      }
+
+      if (arg.find("+tracerv-expected-output=") == 0) {
+        const std::string fname =
+            arg.c_str() + 25; // 25 is the length of the argument to find
+        expected = fopen(fname.c_str(), "w");
```

Comment:
The scope of open files should be narrowed. The constructor can parse the file name, which should only be opened in the write function below.